httpgrpc/protocol_versions.go

```diff
@@ -30,3 +36,24 @@ const (
 	UnaryRpcContentType_V1  = "application/x-protobuf"
 	StreamRpcContentType_V1 = "application/x-httpgrpc-proto+v1"
 )
+
+func getCodec(contentType string) encoding.Codec {
+	if contentType == UnaryRpcContentType_V1 || contentType == StreamRpcContentType_V1 {
+		return encoding.GetCodec(grpcproto.Name)
+	}
+
+	// Try to find the best encoder.
+	if strings.HasPrefix(contentType, "application/grpc-") {
+		if c := encoding.GetCodec(strings.TrimPrefix(contentType, "application/grpc-")); c != nil {
+			return c
+		}
+	}
+
+	if strings.HasPrefix(contentType, "application/") {
```

Comment:
This will be correct for unary RPCs, but not for streaming RPCs. This package uses a custom envelope for messages when streaming that means a streamed response body will not actually be value JSON.

It would be nice, however, if streams that used JSON _were_ valid JSON and just concatenated newline-separate JSON values in the stream, since that can trivially be correctly decoded by a stream consumer.

That would likely require some more intrusive changes to `io.go` so that the framing can vary by codec: no envelope for JSON, but size-prefixed envelope for proto binary.

To that end, it may be better to just add direct support for "application/json" instead of trying to be generic like this.

Reply:
I... don't really understand what you think I should do right now.  I thought the entire point was to remove JSON specific stuff out of this repo in favor of pluggability.

As far as io.go, If we also made the framing pluggable, it seems like we'd need to allow for grpchan-specific configuration and pluggability, as opposed to riding on general grpc codec pluggability.  I can't begin to imagine what that would look like.  (I'm still of the opinion that streaming RPCs should probably just always be websockets anyway, so I'm not sure it's worth going too far down the rabbit hole anyway.)

Reply:
>  I thought the entire point was to remove JSON specific stuff out of this repo

Pluggability is neat, but I certainly would not characterize that as the entire point. IMO, the entire point is JSON support (for our TS clients) in a fashion that is consistent and potentially usable (without concern for backwards compatibility breakages) by non-FullStory consumers of this repo.

The content type currently for streams is "application/x-httpgrpc-proto+v1". If you really want to do generic/pluggable content-types, then we'd want to be able to plug the "proto" part of that. So "application/x-httpgrpc-json+v1" is basically what you've implemented for the streaming part of it.

> I'm still of the opinion that streaming RPCs should probably just always be websockets anyway

That is certainly necessary for full-duplex bidi streaming, but it's not necessary for other kinds of streaming. Also, we already make use of non-web-socket streams (solrmonitor, reflection), so I don't think it's an option to _require_ it, for backwards compatibility reasons. As a data point: the grpc-web protocol does not use web sockets; and I think it would ultimately be valuable to move that direction, so we're using a standard which could open up interesting options in the future for use of stuff from the open-source community. (Admittedly, grpc-web technically only supports server streams; however there is an alternate implementation that supports optional web sockets, which is what I think we'd want to do, too.)

Reply:
> Pluggability is neat, but I certainly would not characterize that as the entire point. IMO, the entire point is JSON support (for our TS clients) in a fashion that is consistent and potentially usable (without concern for backwards compatibility breakages) by non-FullStory consumers of this repo.

Okay, this probably what we should figure out, because I think there are competing forces at play-- not really even between us having different points of view, but us each having a certain degree of internal ambivalence as well. :)

I think I would stack rank my interests like this:

1) JSON over HTTP works correctly inside FS as it does today.

2) FS can follow master in this repo instead of being on a branch.

3) Extending our current end to end solution to support streams over websockets more generally would be nice.  The server piece is currently tucked away in BWAMP, and there's no support in `protoc-gen-typescript` for generating client stubs for streaming over websocket.  There's currently no direct need in product, but I think at some point they'd find it useful.

4) It would be nice if some form of our end-to-end solution were made available to the outside world (including `protoc-gen-typescript`!) but I don't want to change our basic solution to make this happen.  Having the browser be able to JSON parse / JSON serialize typescript objects natively into the wire format, and having the wire format be directly inspectable in browser debuggers, is too good to compromise on.

5) Extending our current end to end solution to support non-duplex streams over plain HTTP would be nice to have, but I don't think we have a burning need for it immediately.  This seems lower priority than websockets since websockets could support any kind of stream.

6) Anything involving compatibility with `grpc-web`. :)

I imagine your list of priorities is probably in a somewhat different order.

> The content type currently for streams is "application/x-httpgrpc-proto+v1". If you really want to do generic/pluggable content-types, then we'd want to be able to plug the "proto" part of that. So "application/x-httpgrpc-json+v1" is basically what you've implemented for the streaming part of it.

I don't, really, at least not as a direct goal.  If you think a hard coded `application/json` solution that lives entirely in this repo is something we can ship, that's fine with me as far as meeting my list of goals, and maybe that's a path to ultimately open sourcing `protoc-gen-typescript` and having an open source end-to-end solution.

Reply:
I think we are pretty strongly aligned so perhaps our differences have been inadvertent miscommunication. I apologize if that's the case.

Mine are definitely a different order, but my number 1 is the same as yours. So let's land something minimal that also gives us your number two. Which, as I stated before, just means being clear about what we support and committing to backwards compatibility. And I think that just means documenting that JSON doesn't support streaming and also preventing it at runtime by failing-fast if the server gets "application/json" for a streaming RPC.

The more I think about the means of pluggability in here now, the more icky it feels to me. For example, the text codec for protos would not really use "application/text" as the content-type. The scheme happens to conveniently work for JSON (since "application/json" is a well-known media type and "json" is a well-known codec name for grpc). But I don't think it really extrapolates cleanly to any arbitrary codec name. So I think the formulation we had before -- pre-rebase -- is a better approach for now.

FWIW, here's my priority ranking:

1. JSON over HTTP works correctly inside FS as it does today.

2. Compatibility with grpc-web. _This actually isn't that hard, would be way more interesting for the open-source community, and it could be interesting for us since it means we could take advantage of future open-source things built around the grpc-web spec. For example: a service mesh that does layer-7 proxying/intercepting would know more about our traffic and provide more interesting features/configuration options if we used a standard like grpc-web._

3. Extending our current end-to-end solution to support streams over websockets.

4. Extending our current end-to-end solution to support non-duplex streams over plain HTTP. _This already "just works" and is in use today (solrmonitor, reflection). But I'm guessing you mean extending this capability to support JSON, in which case I think this is completely moot if we did the above two items._

5. It would be nice if some form of our end-to-end solution were made available to the outside world (including protoc-gen-typescript!). _I don't see a very compelling reason other organizations would adopt much of this it if we don't bother supporting/conforming to standards, like grpc-web. FWIW, there *already* are open-source TS plugins for protoc that have reasonable adoption. The value in writing our own was really that it generated TS code that was backwards-compatible with our previous iteration using Go->TS "tsexport"._

6. FS can follow master in this repo instead of being on a branch. _I'm surprised this is so important to you since you are usually much more willing to fork open-source projects than I am._

Reply:
> I think we are pretty strongly aligned so perhaps our differences have been inadvertent miscommunication. I apologize if that's the case.

No worries, it's very easy to end up accidentally going around in circles.  There's so much in our heads you either have to write a ton of words to get the comprehensive picture down (as we've now done) or sync up face to face.

> Mine are definitely a different order, but my number 1 is the same as yours. So let's land something minimal that also gives us your number two. Which, as I stated before, just means being clear about what we support and committing to backwards compatibility. And I think that just means documenting that JSON doesn't support streaming and also preventing it at runtime by failing-fast if the server gets "application/json" for a streaming RPC.

That seems reasonable, but would the actual JSON encoder be pluggable, or would it be hard coded as it was before to the specific jsonpb marshaler and settings I had in there?

> The more I think about the means of pluggability in here now, the more icky it feels to me. For example, the text codec for protos would not really use "application/text" as the content-type. The scheme happens to conveniently work for JSON (since "application/json" is a well-known media type and "json" is a well-known codec name for grpc). But I don't think it really extrapolates cleanly to any arbitrary codec name. So I think the formulation we had before -- pre-rebase -- is a better approach for now.

Fair.

> 2. Compatibility with grpc-web. _This actually isn't that hard, would be way more interesting for the open-source community, and it could be interesting for us since it means we could take advantage of future open-source things built around the grpc-web spec. For example: a service mesh that does layer-7 proxying/intercepting would know more about our traffic and provide more interesting features/configuration options if we used a standard like grpc-web._

What does this actually mean, really though?  I mean if someone is using grpc-web, why would they need grpchan?  I'm unclear on what that compatibility would buy us, but the biggest thing for me is that that first item on their roadmap is this:

> ## Non-Binary Message Encoding
> 
> The binary protobuf encoding format is not most CPU efficient for browser
> clients. Furthermore, the generated code size increases as the total protobuf
> definition increases.
> 
> For Google's Web applications (e.g. gmail), we use a JSON like format which is
> comparable to JSON in efficiency but also very compact in both the message size
> and code size.

I mean.. they literally don't even have a JSON format defined publicly, despite the fact that they're actually using it internally (because obviously, it's way way better to do so).

> 4. Extending our current end-to-end solution to support non-duplex streams over plain HTTP. _This already "just works" and is in use today (solrmonitor, reflection). But I'm guessing you mean extending this capability to support JSON, in which case I think this is completely moot if we did the above two items._

Yeah, using JSON.  The scope of the whole discussion (for me) is accessing GRPC from a browser.  Server to server can just use the stock protocol.  This is why even server streaming over XHR to a browser is kind of a pain, there's far fewer guarantees about how and when partial responses are delivered and no framing.  Imaging trying to incrementally parse partial content that might end halfway through the next JSON message?  Websockets just offer a ton of advantages here.  But listing it for completeness-- there may be server streaming cases where it's okay to buffer the entire response on the client before you start parsing it.

> 5. It would be nice if some form of our end-to-end solution were made available to the outside world (including protoc-gen-typescript!). _I don't see a very compelling reason other organizations would adopt much of this it if we don't bother supporting/conforming to standards, like grpc-web. FWIW, there _already_ are open-source TS plugins for protoc that have reasonable adoption. The value in writing our own was really that it generated TS code that was backwards-compatible with our previous iteration using Go->TS "tsexport"._

This is only somewhat true; the main thing is that it's compatible with the `jsonpb` marshaller.

> 6. FS can follow master in this repo instead of being on a branch. _I'm surprised this is so important to you since you are usually much more willing to fork open-source projects than I am._

Well I think this explains why this PR's been open for a year 😂 .  You're not highly motivated to have this land in master, and I'm not highly motivated to do much work beyond what we need for FS. 😛